### PR TITLE
api version selection added

### LINKED ISSUE #163: version post 2.8
Hello! I would like to know how to use this library for newer versions of api graph since in the FbBotApp class of this project says

```
    /** 
     * FB Messenger API Url 
     * 
     * @var string 
     */ 
    protected $apiUrl = 'https://graph.facebook.com/v2.8/';
```

and this version of the api is about to cease to be available.

Thank you
Gabriel

### PR DESCRIPTION
**Pull request by o2-web**
Fixes #163 

Summary of changes proposed in this Pull Request:
- add version to __construct
- add protected $baseApiUrl property
- protected $apiVersion property
- set property $apiUrl = null;

PR checklist:
- [x] Updated CHANGELOG.md

### CODE REVIEW

File: FbBotApp.php [hunk 1]
```diff
@@ -44,10 +44,15 @@ class FbBotApp
...
      * FbBotApp constructor.
      * @param string $token
+     * @param string $version optional
      */
-    public function __construct($token)
+    public function __construct($token, $version = null)
     {
         $this->token = $token;
+        
+        if (! is_null($version)) {
```
Comment (wittfabian) [inline FbBotApp.php]: Remove the blank between "!" and  "is_null"

**Comment by wittfabian** [comment]
Great, but we should add a few changes.

Add a base api url:
`protected $baseApiUrl = 'https://graph.facebook.com/';`

Add a base api version:
`protected $baseApiVersion = 'v2.8'`

Use the full string with "v" as version number:
```
if (is_null($version)) {
    $version = $this->$baseApiVersion
}
$this->apiUrl = $this->$baseApiUrl . $version . "/";    
```

What do you think?

**Comment by 02-web** [comment]
proposes to put the default Api Version to the value 'v3.2' .
I think that version 2.8 cannot be left as a default option. Sorry for bad English)

**Comment by wittfabian** [comment]
Can we replace $apiUrl?

**Comment by o2-web** [comment]
> Can we replace $apiUrl?

we can, but I don't really like it. We have to think how to make more beautiful

**Comment by o2-web** [comment]
@wittfabian I suggest initially setting $apiUrl to `null` or an empty string
After all, in any case we define it in the constructor

**Comment by wittfabian** [comment]
Yes, but we should avoid duplicates.

**Comment by o2-web** [comment]
sorry, what kind of duplication are you talking about? $apiUrl?
if so, $token is also duplicated. I meant duplicating the Api version, but not the entire parameter. I am confused, what is the version of the Api being present in the form of a string simultaneously in two properties of $apiVersion & $apiUrl.
Oh, this language barrier apparently)

File: FbBotApp.php [hunk 2]
```diff
@@ -20,13 +20,27 @@ class FbBotApp
...
+    protected $apiVersion = 'v3.2'
 
     /**
      * FB Messenger API Url
      *
-     * @var string
+     * @var null|string
      */
-    protected $apiUrl = 'https://graph.facebook.com/v2.8/';
+    protected $apiUrl;
```
Comment (wittfabian) [inline FbBotApp.php]: Use `protected $apiUrl = null;`

**Comment by wittfabian** [comment]
Are all functions tested with api v3.2?

**Comment by 02-web** [comment]
> Are all functions tested with api v3.2?

Facebook recommends going to 3.2, but there is no documentation of the differences anywhere, they don't work much at all, and hardly anyone or anything will change that. It's just a formality and a report to the employer I think. 

so far as the topic, I constantly test and all is well, even one gets the feeling that nothing has changed

**Comment by o2-web** [comment]
tested on regular and structured messages and all types buttons + action typing. On send location button too

**Comment by wittfabian** [comment]
Last thing:  Updated CHANGELOG.md

**Comment by jason-engage** [comment]
Why are we updating to v3.2? Perhaps its wiser to go to 2.12 (or wtv the max is)

Usually major number update indicates breaking change. Also, there is a change from App level Ids to Page level Ids that I haven't tested, nor heard of anyone discussing... 

I don't feel that Ernest fully understands the possible implications... Can we change from 3.2 to 2.12?

**Comment by wittfabian** [comment]
That's why the pull request has not been merded yet.

**Comment by o2-web** [comment]
> I don't feel that Ernest fully understands the possible implications... Can we change from 3.2 to 2.12?

API Graph v2.8 will be deprecated on Apr 18, 2019. Therefore, version v2.12 and even more cannot be left as default

I changed v3.2 on v2.8
but on the other hand v2.8 also cannot be left as default since it will be deprecated in April

perhaps the best solution would be to put the default version 2.9 or 3.0, but then you need to test all the features in these versions.

**Comment by jason-engage** [comment]
2.12 is ok until 2020.

I suggest you change it to 2.12, otherwise I'm pretty sure it your update won't be merged.

**Comment by wittfabian** [comment]
Perfekt, please update the CHANGELOG.md.

**Comment by wittfabian** [comment]
@jason-engage, do you agree with the changes?

**Comment by jason-engage** [comment]
> @jason-engage, do you agree with the changes?

Yeah I think its fine, code looks ok, and 2.12 should be fine in terms of compatibility. However, I know theres an API version selector in the Facebook App settings, yet I'm not sure what happens if less than 2.12 is selected there while 2.12 is used here. Anyone know?

**Comment by wittfabian** [comment]
Good question...

After the merge we should release v1.2.0

**Comment by jason-engage** [comment]
Is 1.2.0 just a numerical change, or do you have any new features to add?

I do tend to think this script could use an error code processing layer that is triggered with the response from the send call.... Instead of forcing everyone to build their one, the app could have some standard one built in.

I also have a bunch of random (higher-level) functions I've built that I don't mind open sourcing, I'm just thinking how they could be added to this repo without taking it too far in some other directions. For example, we could auto-build a Gallery Card from any FB page post ID, or group of Posts IDs...

**Comment by wittfabian** [comment]
I think the change is big (api v2.8 is deprecated), so 1.1.2 would be the wrong version number.

What does that refer to?
> I do tend to think this script could use an error code processing layer that is triggered with the response from the send call.... Instead of forcing everyone to build their one, the app could have some standard one built in.
> 
> I also have a bunch of random (higher-level) functions I've built that I don't mind open sourcing, I'm just thinking how they could be added to this repo without taking it too far in some other directions. For example, we could auto-build a Gallery Card from any FB page post ID, or group of Posts IDs...

**Comment by jason-engage** [comment]
Oh just some other ideas not referring to this thread! :) 

Graph 2.8 will be out of date in a couple weeks, so a change must be made. I don't think there is a difference between 2.8 and 2.12, except that 2.12 expires in 2020.

I think so far the difference between 2.12 and 3.0 is the App level vs Page level user ids. Not exactly sure but needs to be closely looked at.

**Comment by wittfabian** [comment]
Reference: https://developers.facebook.com/docs/pages/access-tokens/psid-api

**Pull request merged**